src/adapter/src/rbac.rs

```diff
@@ -51,70 +66,115 @@ impl std::fmt::Display for Attribute {
     }
 }
 
-/// Errors that can occur due to an unauthorized action.
 #[derive(Debug)]
-pub struct UnauthorizedError {
-    /// The action that was forbidden.
-    action: String,
-    kind: UnauthorizedErrorKind,
+pub struct Ownership(ObjectId);
+
+impl Ownership {
+    /// Reports whether a role has ownership over an object.
+    fn check_role(&self, role_id: &RoleId, catalog: &impl SessionCatalog) -> bool {
+        match self.0 {
+            ObjectId::Cluster(id) => catalog.get_cluster(id).owner_id() == *role_id,
+            ObjectId::ClusterReplica((cluster_id, replica_id)) => {
+                catalog
+                    .get_cluster_replica(cluster_id, replica_id)
+                    .owner_id()
+                    == *role_id
+            }
+            ObjectId::Database(id) => catalog.get_database(&id).owner_id() == *role_id,
+            ObjectId::Schema((database_id, schema_id)) => {
+                catalog
+                    .get_schema(
+                        &ResolvedDatabaseSpecifier::Id(database_id),
+                        &SchemaSpecifier::Id(schema_id),
+                    )
+                    .owner_id()
+                    == *role_id
+            }
+            ObjectId::Item(id) => catalog.get_item(&id).owner_id() == *role_id,
+        }
+    }
+
+    /// Reports whether any role has ownership over an object.
+    fn check_roles(&self, role_ids: &BTreeSet<RoleId>, catalog: &impl SessionCatalog) -> bool {
+        role_ids
+            .iter()
+            .any(|role_id| self.check_role(role_id, catalog))
+    }
 }
 
+/// Errors that can occur due to an unauthorized action.
 #[derive(Debug)]
-enum UnauthorizedErrorKind {
+pub enum UnauthorizedError {
     /// The action can only be performed by a superuser.
-    Superuser,
+    Superuser { action: String },
     /// The action requires a specific attribute.
-    Attribute(Attribute),
+    Attribute {
+        action: String,
+        attribute: Attribute,
+    },
+    /// The action requires ownership of an object.
+    Ownership { objects: Vec<(ObjectType, String)> },
     /// The action requires one or more privileges.
-    Privilege { reason: Option<String> },
-}
-
-impl fmt::Display for UnauthorizedError {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        write!(f, "permission denied to {}", self.action)
-    }
+    Privilege {
+        action: String,
+        reason: Option<String>,
+    },
 }
 
 impl UnauthorizedError {
     pub fn superuser(action: String) -> UnauthorizedError {
```

Comment:
Is there a reason we're maintaining these constructors when we also allow creating the variants directly? In this case I'd say we should just create the variants directly

Reply:
Good point, I've removed the constructors.